File: scripts/fix-missing-tracks.mjs
import { readFile, writeFile } from "node:fs/promises";
import path from "node:path";

const root = process.cwd();
const catalogPath = path.join(root, "src/data/catalog.generated.json");
const delayMs = Number(process.env.APPLE_DELAY_MS ?? 2000);
const requestTimeoutMs = 20000;

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function slugify(value) {
  return value
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .replace(/&/g, " and ")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 72);
}

function normalizeText(value) {
  return String(value ?? "")
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/\b(the|a|an)\b/g, "")
    .replace(/[^a-z0-9]+/gi, " ")
    .toLowerCase()
    .trim();
}

function similarity(a, b) {
  const left = new Set(normalizeText(a).split(/\s+/).filter(Boolean));
  const right = new Set(normalizeText(b).split(/\s+/).filter(Boolean));
  if (!left.size || !right.size) return 0;
  let overlap = 0;
  for (const token of left) {
    if (right.has(token)) overlap += 1;
  }
  return overlap / Math.max(left.size, right.size);
}

async function fetchJson(url) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), requestTimeoutMs);
  try {
    const response = await fetch(url, {
      headers: { "User-Agent": "AlbumVault/0.1 fixer" },
      signal: controller.signal
    });
    if (!response.ok) throw new Error(`${response.status}: ${response.statusText}`);
    return response.json();
  } finally {
    clearTimeout(timeout);
  }
}

async function lookupTracks(collectionId) {
  const url = `https://itunes.apple.com/lookup?id=${collectionId}&entity=song&country=US`;
  const data = await fetchJson(url);
  return (data.results ?? [])
    .filter((item) => item.wrapperType === "track" && item.kind === "song")
    .sort((a, b) => (a.discNumber ?? 1) - (b.discNumber ?? 1) || (a.trackNumber ?? 0) - (b.trackNumber ?? 0))
    .map((track) => ({
      discNumber: track.discNumber ?? 1,
      trackNumber: track.trackNumber ?? 0,
      title: track.trackName,
      durationMs: track.trackTimeMillis ?? null,
      previewUrl: track.previewUrl ?? null
    }));
}

function scoreMatch(album, candidate) {
  const titleScore = similarity(album.title, candidate.collectionName);
  const artistScore = similarity(album.artist, candidate.artistName);
  const year = candidate.releaseDate ? Number(candidate.releaseDate.slice(0, 4)) : undefined;
  const yearScore = year === album.year ? 1 : year && Math.abs(year - album.year) <= 1 ? 0.6 : 0;
  return titleScore * 0.48 + artistScore * 0.42 + yearScore * 0.1;
}

async function searchApple(album) {
  const term = encodeURIComponent(`${album.artist} ${album.title}`);
  const urls = [
    `https://itunes.apple.com/search?term=${term}&media=music&entity=album&attribute=albumTerm&limit=10&country=US`,
    `https://itunes.apple.com/search?term=${term}&media=music&entity=album&limit=15&country=US`,
    `https://itunes.apple.com/search?term=${term}&media=music&entity=album&limit=15&country=GB`
  ];
  const ranked = [];
  for (const url of urls) {
    try {
      const data = await fetchJson(url);
      ranked.push(
        ...(data.results ?? [])
          .filter((c) => c.wrapperType === "collection")
          .map((c) => ({ candidate: c, score: scoreMatch(album, c) }))
      );
    } catch (e) {
      console.log(`    ⚠ Search failed for ${album.title}: ${e.message}`);
    }
    await sleep(Math.round(delayMs / 3));
  }
  ranked.sort((a, b) => b.score - a.score);
  return ranked[0]?.score >= 0.5 ? ranked[0] : null;
}

async function main() {
  const catalog = JSON.parse(await readFile(catalogPath, "utf8"));
  const empty = catalog.albums.filter((a) => a.tracks.length === 0);
  console.log(`Found ${empty.length} albums with 0 tracks\n`);

  let fixed = 0;
  let failed = 0;

  for (const album of empty) {
    console.log(`🎵 #${album.rank} "${album.title}" by ${album.artist}`);

    if (album.appleCollectionId) {
      // Has ID but no tracks — just lookup tracks
      console.log(`  → Has collectionId ${album.appleCollectionId}, looking up tracks...`);
      try {
        album.tracks = await lookupTracks(album.appleCollectionId);
        await sleep(delayMs);
        console.log(`  ✅ Got ${album.tracks.length} tracks`);
        fixed++;
      } catch (e) {
        console.log(`  ❌ Track lookup failed: ${e.message}`);
        failed++;
      }
    } else {
      // No collectionId — search Apple Music
      console.log(`  → No collectionId, searching Apple Music...`);
      try {
        const match = await searchApple(album);
        await sleep(delayMs);
        if (match) {
          album.appleCollectionId = match.candidate.collectionId;
          album.appleCollectionUrl = match.candidate.collectionViewUrl ?? null;
          album.appleArtworkUrl = match.candidate.artworkUrl100?.replace(/\/\d+x\d+bb\./, "/600x600bb.") ?? null;
          if (!album.genre) album.genre = match.candidate.primaryGenreName ?? null;
          
          console.log(`  → Matched: ${match.candidate.collectionName} (score: ${match.score.toFixed(3)}), looking up tracks...`);
          album.tracks = await lookupTracks(match.candidate.collectionId);
          await sleep(delayMs);
          console.log(`  ✅ Got ${album.tracks.length} tracks + collectionId=${album.appleCollectionId}`);
          fixed++;
        } else {
          console.log(`  ❌ No good match found (${ranked.length} candidates checked)`);
          failed++;
        }
      } catch (e) {
        console.log(`  ❌ Search failed: ${e.message}`);
        failed++;
      }
    }
  }

  // Save updated catalog
  catalog.metadata.generatedAt = new Date().toISOString();
  await writeFile(catalogPath, JSON.stringify(catalog, null, 2));
  
  const stillEmpty = catalog.albums.filter((a) => a.tracks.length === 0);
  console.log(`\n🏁 Done. Fixed: ${fixed}, Failed: ${failed}, Still empty: ${stillEmpty.length}`);
  if (stillEmpty.length > 0) {
    console.log("Still missing:");
    for (const a of stillEmpty) {
      console.log(`  #${a.rank} "${a.title}" by ${a.artist}`);
    }
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: scripts/fix-missing-tracks.py
#!/usr/bin/env python3
"""Comprehensive Apple Music track fixer."""
import json, time, urllib.request, urllib.parse, socket, sys

socket.setdefaulttimeout(15)
CATALOG_PATH = "src/data/catalog.generated.json"
DELAY = 1.0

def fetch(url):
    req = urllib.request.Request(url, headers={"User-Agent": "AlbumVault/0.1"})
    with urllib.request.urlopen(req, timeout=15) as r:
        return json.loads(r.read())

def lookup_tracks(cid):
    for c in ["US","GB","JP","DE","NL","FR","AU","NZ"]:
        try:
            data = fetch(f"https://itunes.apple.com/lookup?id={cid}&entity=song&country={c}")
            tracks = []
            for r in data.get("results", []):
                if r.get("wrapperType") == "track" and r.get("kind") == "song":
                    tracks.append({
                        "discNumber": r.get("discNumber", 1),
                        "trackNumber": r.get("trackNumber", 0),
                        "title": r.get("trackName", ""),
                        "durationMs": r.get("trackTimeMillis"),
                        "previewUrl": r.get("previewUrl")
                    })
            if tracks:
                tracks.sort(key=lambda t: (t["discNumber"], t["trackNumber"]))
                return tracks, c
        except:
            pass
        time.sleep(0.25)
    return [], None

def find_via_track(artist, track_name):
    term = urllib.parse.quote(f"{track_name} {artist}")
    art_low = artist.lower()
    for country in ["US","GB","JP","DE"]:
        try:
            data = fetch(f"https://itunes.apple.com/search?term={term}&media=music&entity=song&limit=10&country={country}")
            for r in data.get("results", []):
                if r.get("kind") != "song": continue
                aname = r.get("artistName","").lower()
                if art_low not in aname: continue
                cname = r.get("collectionName","").lower()
                if any(x in cname for x in ["single","ep","remix","compilation","greatest","tribute",": solo "]):
                    continue
                return r.get("collectionId"), r.get("collectionName"), country
        except: pass
        time.sleep(DELAY/3)
    return None,None,None

def find_album(artist, title):
    term = urllib.parse.quote(f"{artist} {title}")
    art_low = artist.lower()
    tit_low = title.lower()
    for country in ["US","GB","JP","DE","NL","FR","AU"]:
        try:
            data = fetch(f"https://itunes.apple.com/search?term={term}&media=music&entity=album&limit=15&country={country}")
            for r in data.get("results", []):
                if r.get("wrapperType") != "collection": continue
                aname = r.get("artistName","").lower()
                if art_low not in aname: continue
                cname = r.get("collectionName","").lower()
                if any(x in cname for x in ["single","ep","remix","karaoke","deluxe","expanded","anniversary","tribute","lullaby",": solo","piano tribute"]):
                    continue
                if tit_low == cname or (len(tit_low)>8 and tit_low in cname):
                    return r, country
        except: pass
        time.sleep(DELAY/3)
    return None,None

TRACK_HINTS = {
    "Nevermind":"Smells Like Teen Spirit","Exile on Main St.":"Tumbling Dice",
    "It Takes a Nation of Millions to Hold Us Back":"Bring the Noise",
    "Lemonade":"Formation","The Chronic":"Nuthin' But a G Thang","Blonde":"Nikes",
    "Exile in Guyville":"Fuck and Run","Appetite for Destruction":"Sweet Child O' Mine",
    "Reasonable Doubt":"Can't Knock the Hustle","Plastic Ono Band":"Mother",
    "Automatic for the People":"Everybody Hurts","Live Through This":"Doll Parts",
    "When the Pawn...":"Paper Bag","Disintegration":"Pictures of You",
    "The Downward Spiral":"Closer","Paul's Boutique":"Shake Your Rump",
    "1999":"Little Red Corvette","Maggot Brain":"Maggot Brain",
    "The Marshall Mathers LP":"The Real Slim Shady","Parallel Lines":"Heart of Glass",
    "Channel Orange":"Thinkin Bout You","A Love Supreme":"Acknowledgement",
    "Mama's Gun":"Didn't Cha Know","Different Class":"Common People",
    "20 Golden Greats":"That'll Be the Day","Fear of a Black Planet":"Fight the Power",
    "Blood Sugar Sex Magik":"Under the Bridge","Electric Warrior":"Get It On",
    "Slanted and Enchanted":"Summer Babe","Diamond Life":"Smooth Operator",
    "Midnight Marauders":"Award Tour","Homogenic":"J\u00f3ga","Pink Moon":"Pink Moon",
    "Rage Against the Machine":"Killing in the Name","Raising Hell":"Walk This Way",
    "Yankee Hotel Foxtrot":"Jesus, Etc.","Hospice":"Kettering",
    "Mama Said Knock You Out":"Mama Said Knock You Out","American Idiot":"American Idiot",
}

HARDCODED = {
    "Trout Mask Replica": [
        "Frownland","The Dust Blows Forward 'n the Dust Blows Back","Dachau Blues",
        "Ella Guru","Hair Pie: Bake 1","Moonlight on Vermont","Pachuco Cadaver",
        "Bill's Corpse","Sweet Sweet Bulbs","Neon Meate Dream of a Octafish",
        "China Pig","My Human Gets Me Blues","Dali's Car","Hair Pie: Bake 2",
        "Pena","Well","When Big Joan Sets Up","Fallin' Ditch","Sugar 'n Spikes",
        "Ant Man Bee","Orange Claw Hammer","Wild Life","She's Too Much for My Mirror",
        "Hobo Chang Ba","The Blimp (mousetrapreplica)","Steal Softly Thru Snow",
        "Old Fart at Play","Veteran's Day Poppy",
    ],
}

def make_tracks(names):
    return [{"discNumber":1,"trackNumber":i+1,"title":n,"durationMs":None,"previewUrl":None} for i,n in enumerate(names)]

def main():
    with open(CATALOG_PATH) as f:
        catalog = json.load(f)
    empty = [a for a in catalog["albums"] if len(a.get("tracks",[])) == 0]
    if not empty:
        print("All albums have tracks!")
        return
    print(f"{len(empty)} albums with 0 tracks\n")
    fixed = 0; failed = 0; total = len(empty)
    
    for i, album in enumerate(empty):
        ttl = album["title"]; art = album["artist"]; rnk = album["rank"]; cid = album.get("appleCollectionId")
        pfx = f"[{i+1}/{total}] #{rnk} \"{ttl}\" by {art}"
        print(pfx, end=" ", flush=True)
        
        if ttl in HARDCODED:
            album["tracks"] = make_tracks(HARDCODED[ttl])
            print(f"\u2192 \u2705 {len(album['tracks'])} tracks (hardcoded)"); fixed += 1; continue
        
        if cid:
            tracks, store = lookup_tracks(cid)
            if tracks:
                album["tracks"] = tracks
                print(f"\u2192 \u2705 {len(tracks)} tracks (existing id, {store})"); fixed += 1; time.sleep(DELAY); continue
        
        hint = TRACK_HINTS.get(ttl)
        if hint:
            t_cid, t_name, t_country = find_via_track(art, hint)
            if t_cid:
                tracks, tstore = lookup_tracks(t_cid)
                if tracks:
                    album["tracks"] = tracks; album["appleCollectionId"] = t_cid
                    album["appleCollectionUrl"] = f"https://music.apple.com/album/{t_cid}"
                    print(f"\u2192 \u2705 {len(tracks)} tracks (via \"{hint}\", id={t_cid})"); fixed += 1; time.sleep(DELAY); continue
        time.sleep(DELAY)
        
        match, m_country = find_album(art, ttl)
        if match:
            m_cid = match.get("collectionId")
            if m_cid and m_cid != cid:
                tracks, mstore = lookup_tracks(m_cid)
                if tracks:
                    album["tracks"] = tracks; album["appleCollectionId"] = m_cid
                    album["appleCollectionUrl"] = match.get("collectionViewUrl")
                    aw = match.get("artworkUrl100","")
                    if aw: album["appleArtworkUrl"] = aw.replace("100x100bb","600x600bb")
                    if not album.get("genre"): album["genre"] = match.get("primaryGenreName")
                    print(f"\u2192 \u2705 {len(tracks)} tracks (direct, id={m_cid})"); fixed += 1; time.sleep(DELAY); continue
        
        print(f"\u2192 \u274c Not found"); failed += 1
    
    catalog["metadata"]["generatedAt"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    with open(CATALOG_PATH, "w") as f:
        json.dump(catalog, f, indent=2); f.write("\n")
    
    still = [a for a in catalog["albums"] if len(a.get("tracks",[])) == 0]
    print(f"\n{'='*60}")
    print(f"Fixed: {fixed} | Failed: {failed} | Still empty: {len(still)}/{total}")
    if still:
        for a in still:
            print(f"  #{a['rank']} \"{a['title']}\" by {a['artist']}")

if __name__ == "__main__":
    main()
